### ISSUE TITLE
[TBAA] Miscompilation at O2/3

### ISSUE THREAD

**Issue by dtcxzyw** [llvm/llvm-project#198847]
Reproducer:

```c
#include <stdint.h>
#include <stdio.h>

#define a(b, c)                                                                \
  c builtin_##b(c d, c p) { return (d > p) - (d < p); }
a(scmp, int) uint64_t e;
int32_t f[] = {2, 2, 12};
int32_t g;
int32_t h[4];
static int32_t *i = &h[2];
uint64_t t;
int32_t *j = &h[2];
uint64_t *k = &t;
uint64_t **l = &k;
uint64_t ***m = &l;
uint64_t n(int32_t *, int16_t);
uint8_t o() {
  n(&f[1], e);
  return 0;
}
uint64_t n(int32_t *q, int16_t r) {
  int32_t **s = &i;
  for (g = 2; g; g--)
    **s = (***m = builtin_scmp(*j, 7), f[g]);
  return 0;
}
int main() {
  o();
  printf("%d\n", (int)t);
  return 0;
}
```

```
> ../../llvm-build/bin/clang -O2 test.c && ./a.out
-1
> ../../llvm-build/bin/clang -O2 test.c -fno-strict-aliasing && ./a.out
-1
> ../../llvm-build/bin/clang -O0 test.c && ./a.out
1
```

I failed to use llubi to obtain IR repro. The reduced result is invalid (loading i32 from a ptr global).
Original csmith reproducer: https://gist.github.com/dtcxzyw/a78d02bb4134e440ee8485f1de2ac22e
LLVM version: 0f74c51841823ea9946983273484d3ff87369f97

This bug was reported by https://github.com/dtcxzyw/llvm-fuzz-service/issues/185. But it is not a regression caused by https://github.com/llvm/llvm-project/pull/198114.

**Comment by lijinpei-amd** [comment]
Fixed by https://github.com/llvm/llvm-project/pull/198238
